document-legacy/src/document.rs

```diff
@@ -559,50 +621,156 @@ impl Document {
 				let insert_index = folder.layer_ids.iter().position(|&id| id == selected_id).unwrap_or(0) as isize + 1;
 
 				if let Some(new_layer_id) = folder.add_layer(layer, None, insert_index) {
-					let new_path = [folder_path, &[new_layer_id]].concat();
+					let new_path: &Vec<u64> = &([folder_path, &[new_layer_id]].concat());
+
+					let mut responses: Vec<DocumentResponse> = vec![];
+					responses.extend([
+						DocumentChanged,
+						CreatedLayer {
+							path: new_path.to_vec(),
+							is_selected: true,
+						},
+						FolderChanged { path: folder_path.to_vec() },
+					]);
+					responses.extend(update_thumbnails_upstream(path.as_slice()));
+
+					if layer_is_folder {
+						let mut old_to_new_layer_id: HashMap<LayerId, LayerId> = HashMap::new();
+						let new_folder = self.folder_mut(new_path)?;
+						// Clear the new folders layer_ids/layers because they contain the layer_ids/layers of the layer were duplicated
+						// Generate a new next assignment ID to avoid collision
+						new_folder.layer_ids = vec![];
+						new_folder.layers = vec![];
+						new_folder.generate_new_folder_ids();
+
+						for (counter, duplicate_layer) in duplicated_layers.into_iter().enumerate() {
+							let old_layer_id = *duplicate_layer.last().unwrap_or(&0);
+
+							// Iterate through each ID of the current duplicate layer
+							// If the dictionary contains the ID, we know the duplicate folder has been created already. Use the existing layer ID instead of creating a new one
+							let sub_layer = &duplicate_layer[new_path.len()..];
+							let new_sub_path: Vec<u64> = sub_layer.iter().filter_map(|id| old_to_new_layer_id.get(id).cloned()).collect();
+
+							// Combine the new path with the IDs of the duplicate layer path to create the path where we insert the duplicate layer
+							let updated_layer_ref: &Layer = duplicate_layer_objects_sorted.get(counter).unwrap();
+							let mut updated_layer = updated_layer_ref.clone();
+							let updated_layer_path_parent: Vec<u64> = [new_path.to_vec(), new_sub_path].concat();
+
+							// Clear the new folder's layer_ids and layers because they contain the layer_ids/layers of the layer were duplicated
+							if self.is_folder(duplicate_layer) {
+								let updated_layer_as_folder: &mut FolderLayer = updated_layer.as_folder_mut()?;
+								updated_layer_as_folder.layer_ids = vec![];
+								updated_layer_as_folder.layers = vec![];
+								updated_layer_as_folder.generate_new_folder_ids()
+							}
+
+							let result = self
+								.handle_operation(Operation::InsertLayer {
+									layer: Box::new(updated_layer),
+									destination_path: updated_layer_path_parent,
+									insert_index: -1,
+									duplicating: true,
+								})
+								.ok()
+								.flatten()
+								.unwrap_or_default();
+
+							// Collect the new ID of the duplicated layer from the InsertLayer Operation
+							// Map the layer ID of the layer were duplicating to the new ID
```

Comment:
all the comments really helped navigate what you were doing here, thanks

Reply:
nice find